### PR TITLE
feat!: Hugrs now keep a `ExtensionRegistry` with their requirements

### LINKED ISSUE #1613: feat(core): Put extension declarations in top level of `Hugr`
To allow validation of individual HUGRs with any root, without resorting to packages, which only currently allow modules. This would then mean packages would only need to contain modules (no extensions) and can be used mostly for linking. This also aligns well with the model work so should be done as a follow up to that. See #1433.

### PR DESCRIPTION
**Pull request by aborgna-q**
Closes #1613. Depends on #1739.

Hugrs now keep an `Extensions` registry that is automatically computed when using the builder or deserializing from json (using the new `Hugr::load_json`).
This set can contain unneeded extensions, but it should always be sufficient to validate the HUGR definition. Note that this is **not** runtime extensions (see #426, #1734).

A big chunk of the diff is removing the extension registry when finishing building a hugr. The extension tracking is now done automatically while adding operations. 

drive-by: Remove unneeded `set_num_ports` call in `insert_hugr_internal`.

BREAKING CHANGE: Removed `update_validate`. The hugr extensions should be resolved at load time, so we can use `validate` instead.
BREAKING CHANGE: The builder `finish_hugr` function family no longer takes a registry as parameter, and the `_prelude` variants have been removed.

### CODE REVIEW

**Review by ss2165 — COMMENTED** [review]
Main query is why we still need extensions stored at the top level in the package

File: hugr-cli/src/lib.rs [hunk 1]
```diff
@@ -120,13 +119,33 @@ impl AsRef<[Hugr]> for PackageOrHugr {
...
+    pub fn get_package_or_hugr(
+        &mut self,
+        extensions: &ExtensionRegistry,
+    ) -> Result<PackageOrHugr, CliError> {
+        // We need to read the input twice; once to try to load it as a HUGR, and if that fails, as a package.
+        // If `input` is a file, we can reuse the reader by seeking back to the start.
+        // Else, we need to read the file into a buffer.
+        trait SeekRead: Seek + Read {}
+        impl<T: Seek + Read> SeekRead for T {}
+
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply hugr-cli/src/lib.rs:131]: I need to `Read` and `Seek` either a `clap::Input` or a `Cursor`. This is a workaround for `Box<dyn Seek + Read>` not being valid rust.
But in insight I can just define a generic aux function with the logic.

File: hugr-core/src/hugr.rs [hunk 2]
```diff
@@ -391,53 +425,46 @@ mod test {
...
-        let mut hugr: Hugr = serde_json::from_reader(BufReader::new(
-            File::open(test_file!("hugr-1.json")).unwrap(),
-        ))
-        .unwrap();
-        assert!(hugr.update_validate(&PRELUDE_REGISTRY).is_ok());
+        let hugr = Hugr::load_json(
+            BufReader::new(File::open(test_file!("hugr-1.json")).unwrap()),
+            &PRELUDE_REGISTRY,
+        );
+        assert_matches!(&hugr, Ok(_));
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply hugr-core/src/hugr.rs:445]: We're avoiding unnecessary match **guards**. I.e.
```rust
assert_matches!(x, Ok(y) => assert!(f(y)));
```
But this line here is the same as `assert!(hugr.is_ok())` just with a better error msg.

File: hugr-core/src/hugr/validate.rs [hunk 3]
```diff
@@ -28,33 +27,28 @@ use super::ExtensionError;
...
 impl Hugr {
     /// Check the validity of the HUGR, assuming that it has no open extension
     /// variables.
     /// TODO: Add a version of validation which allows for open extension
     /// variables (see github issue #457)
-    pub fn validate(&self, extension_registry: &ExtensionRegistry) -> Result<(), ValidationError> {
-        self.validate_no_extensions(extension_registry)?;
+    pub fn validate(&self) -> Result<(), ValidationError> {
+        self.validate_no_extensions()?;
         #[cfg(feature = "extension_inference")]
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply hugr-core/src/hugr/validate.rs:43]: It's nicer do `if cfg!(...)` since the call gets compiled (and dead-code eliminated), so any syntax error gets reported even when the feature is disabled.
I'll change it here.

File: hugr-passes/src/const_fold.rs [hunk 4]
```diff
@@ -51,40 +51,34 @@ impl ConstantFoldPass {
...
-                    for rem in removes {
-                        // We are optimistically applying these [RemoveLoadConstant] and
-                        // [RemoveConst] rewrites without checking whether the nodes
-                        // they attempt to remove have remaining uses. If they do, then
-                        // the rewrite fails and we move on.
-                        if let Ok(const_node) = hugr.apply_rewrite(rem) {
-                            // if the LoadConst was removed, try removing the Const too.
-                            let _ = hugr.apply_rewrite(RemoveConst(const_node));
-                        }
+    pub fn run<H: HugrMut>(&self, hugr: &mut H) -> Result<(), ConstFoldError> {
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply hugr-passes/src/const_fold.rs:54]: Yeah, the small change made the formatter change the indentation of everything -.-

File: hugr-core/src/package.rs [hunk 5]
```diff
@@ -1,41 +1,43 @@
...
+/// Package of module HUGRs.
 pub struct Package {
     /// Module HUGRs included in the package.
     pub modules: Vec<Hugr>,
-    /// Extensions to validate against.
-    pub extensions: Vec<Arc<Extension>>,
+    /// Extensions used in the modules.
+    ///
+    /// This is a superset of the extensions used in the modules.
+    pub extensions: ExtensionRegistry,
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply hugr-core/src/package.rs:24]: Good question!

This PR does **not** change the serialization format, it's purely change to the runtime representation of hugrs.

In our current serialization model, packages can embed arbitrary extension definitions in the json.
Since the modules do not necessarily use those, if we didn't have this field we would lose track of them.

We could update the serialization and (optionally) include modules in the serialized hugr, but I think we should rather wait for `hugr-model` first.

File: hugr-core/src/ops/validate.rs
Comment (aborgna-q) [inline hugr-core/src/ops/validate.rs:1]: Here I readded a breaking drive-by from further down the PR stack.
It cleans up a validation error message.

**Review by ss2165 — APPROVED** [review]
:shipit:
